extmod/uasyncio.py

```diff
@@ -0,0 +1,565 @@
+# General streams
+
+# Queue and poller for stream IO
+class IOQueue:
+    def __init__(self):
+        self.poller = select.poll()
+        self.map = {}
+    def _queue(self, s, idx):
+        if id(s) not in self.map:
+            entry = [None, None, s]
+            entry[idx] = cur_task
+            self.map[id(s)] = entry
+            self.poller.register(s, select.POLLIN if idx == 0 else select.POLLOUT)
+        else:
+            sm = self.map[id(s)]
+            assert sm[idx] is None
+            assert sm[1 - idx] is not None
+            sm[idx] = cur_task
+            self.poller.modify(s, select.POLLIN | select.POLLOUT)
+        # Link task to this IOQueue so it can be removed if needed
+        cur_task.data = self
+    def _dequeue(self, s):
+        del self.map[id(s)]
+        self.poller.unregister(s)
+    def queue_read(self, s):
+        self._queue(s, 0)
+    def queue_write(self, s):
+        self._queue(s, 1)
+    def remove(self, task):
+        while True:
+            del_s = None
+            for k in self.map: # Iterate without allocating on the heap
+                q0, q1, s = self.map[k]
+                if q0 is task or q1 is task:
+                    del_s = s
+                    break
+            if del_s is not None:
+                self._dequeue(s)
+            else:
+                break
+    def wait_io_event(self, dt):
+        for s, ev in self.poller.ipoll(dt):
+            sm = self.map[id(s)]
+            err = ev & ~(select.POLLIN | select.POLLOUT)
+            #print('poll', s, sm, ev, err)
+            if ev & select.POLLIN or (err and sm[0] is not None):
+                _queue.push_head(sm[0])
+                sm[0] = None
+            if ev & select.POLLOUT or (err and sm[1] is not None):
+                _queue.push_head(sm[1])
+                sm[1] = None
+            if sm[0] is None and sm[1] is None:
+                self._dequeue(s)
+            elif sm[0] is None:
+                self.poller.modify(s, select.POLLOUT)
+            else:
+                self.poller.modify(s, select.POLLIN)
+
+class Stream:
+    def __init__(self, s, e={}):
+        self.s = s
+        self.e = e
+        self.out_buf = b''
+    def get_extra_info(self, v):
+        return self.e[v]
+    async def __aenter__(self):
+        return self
+    async def __aexit__(self, exc_type, exc, tb):
+        await self.close()
+    def close(self):
+        pass
+    async def wait_closed(self):
+        # TODO yield?
+        self.s.close()
+    async def read(self, n):
+        yield _io_queue.queue_read(self.s)
+        return self.s.read(n)
+    async def readline(self):
+        l = b''
+        while True:
+            yield _io_queue.queue_read(self.s)
+            l2 = self.s.readline() # may do multiple reads but won't block
+            l += l2
+            if not l2 or l[-1] == 10: # \n (check l in case l2 is str)
+                return l
+    def write(self, buf):
+        self.out_buf += buf
```

Comment:
Is there any chance to do some eager writing to `self.s` if it's ready?

Reply:
Yes there is, although it'd complicate the function a bit.  Note that with the current implementation there is no heap allocation in this function if `drain` is always called after `write`, because `b'' + buf` is optimised to just return `buf`.